code/xcvm/cosmwasm/contracts/interpreter/src/contract.rs

```diff
@@ -0,0 +1,428 @@
+			Instruction::Transfer { to, assets } =>
+				interpret_transfer(&mut deps, &env, to, assets, response)?,
+		};
+	}
+
+	Ok(response.add_event(Event::new("xcvm.interpreter.executed").add_attribute(
+		"program",
+		core::str::from_utf8(&program.tag).map_err(|_| ContractError::InvalidProgramTag)?,
+	)))
+}
+
+pub fn interpret_call(encoded: Vec<u8>, response: Response) -> Result<Response, ContractError> {
+	let cosmos_msg: CosmosMsg =
+		serde_json_wasm::from_slice(&encoded).map_err(|_| ContractError::InvalidCallPayload)?;
+
+	Ok(response.add_message(cosmos_msg))
+}
+
+pub fn interpret_spawn(
+	deps: &DepsMut,
+	env: &Env,
+	network: NetworkId,
+	salt: Vec<u8>,
+	assets: Funds,
+	program: XCVMProgram,
+	mut response: Response,
+) -> Result<Response, ContractError> {
+	#[derive(Serialize)]
+	struct SpawnEvent {
+		network: NetworkId,
+		salt: Vec<u8>,
+		assets: Funds<Displayed<u128>>,
+		program: XCVMProgram,
+	}
+
+	let config = CONFIG.load(deps.storage)?;
+	let registry_addr = config.registry_address.into_string();
+	let mut normalized_funds = Funds::<Displayed<u128>>::empty();
+
+	for (asset_id, amount) in assets.0 {
+		if amount.is_zero() {
+			// We ignore zero amounts
+			continue
+		}
+
+		let amount = if amount.slope.0 == 0 {
+			// No need to get balance from cw20 contract
+			amount.intercept
+		} else {
+			let query_msg = AssetRegistryQueryMsg::GetAssetContract(asset_id.into());
+
+			let cw20_address: GetAssetContractResponse = deps.querier.query(
+				&WasmQuery::Smart {
+					contract_addr: registry_addr.clone(),
+					msg: to_binary(&query_msg)?,
+				}
+				.into(),
+			)?;
+			let response =
+				deps.querier.query::<BalanceResponse>(&QueryRequest::Wasm(WasmQuery::Smart {
+					contract_addr: cw20_address.addr.clone().into_string(),
+					msg: to_binary(&Cw20QueryMsg::Balance {
+						address: env.contract.address.clone().into_string(),
+					})?,
+				}))?;
+			amount.apply(response.balance.into()).into()
+		};
+
+		if amount.0 > 0 {
+			normalized_funds.0.insert(asset_id, amount.into());
+		}
+	}
+
+	// TODO(probably call the router via a Cw20 `send` to spawn the program and do w/e required with
+	// the funds)
+	for (asset_id, amount) in normalized_funds.clone().0 {
+		let query_msg = AssetRegistryQueryMsg::GetAssetContract(asset_id.into());
+		let cw20_address: GetAssetContractResponse = deps.querier.query(
+			&WasmQuery::Smart { contract_addr: registry_addr.clone(), msg: to_binary(&query_msg)? }
+				.into(),
+		)?;
+		let contract = Cw20Contract(cw20_address.addr);
+		response =
+			response.add_message(contract.call(Cw20ExecuteMsg::Burn { amount: amount.0.into() })?);
+	}
+
+	let data = SpawnEvent { network, salt, assets: normalized_funds, program };
+
+	Ok(response.add_event(
+		Event::new("xcvm.interpreter.spawn")
+			.add_attribute(
+				"origin_network_id",
+				serde_json_wasm::to_string(&config.network_id.0)
+					.map_err(|_| ContractError::DataSerializationError)?,
+			)
+			.add_attribute(
+				"origin_user_id",
```

Comment:
here we do not glue user and network. add 2 attributes. while in interpreter we do. feels incosistent.

Reply:
Can you show me the part where we glue them?

Reply:
Good point but as I mentioned in the other comment, this is exposed to the relayer/frontend (outer world you can say). But the one we use them together is only for internal use of router. So that is the most performant way of doing that. Because we don't have to iterate for two different attributes. But we find one and just decode the raw data. All of the exposed parts use separate attributes for these. So it is consistent.